app/src/main/java/at/aau/serg/activities/MainActivity.kt

```diff
@@ -19,6 +28,48 @@ class MainActivity : AppCompatActivity() {
             v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
             insets
         }
+
+        if(!Authentication(HttpClient(getString(R.string.api_url))).tokenValid(CallbackCreator().createCallback(::startLoginActivity, ::checkIfUserIsAuthenticated), StoreToken(this, Secret()))){
+            this.startActivity(Intent(this, LoginActivity::class.java))
+        }
+    }
+
+    private fun startLoginActivity(){
+        this@MainActivity.startActivity(Intent(this@MainActivity, LoginActivity::class.java))
+    }
+
+    private fun checkIfUserIsAuthenticated(response: Response){
+        if(response.isSuccessful)
+            return
+
+        if(response.code != 403){
+            this@MainActivity.startActivity(Intent(this@MainActivity, LoginActivity::class.java))
+            return
+        }
+
+        if(!Authentication(HttpClient(getString(R.string.api_url))).updateToken(CallbackCreator().createCallback(::startLoginActivity, ::checkIfUpdateAccessTokenWorked), StoreToken(this@MainActivity, Secret()))) {
```

Comment:
Do we create a new instance of HttpClient every time we call a endpoint?
If so is this necessary?
Same with Authentication.

Reply:
Yes, we had created a new instance every time.
I have now changed it to a singleton